### PR TITLE
First try on an implementation for a Docker Genact plugin

### PR DESCRIPTION
**Pull request by Kovah**
I just tried to implement a Docker plugin based on #64.
It basically works but I couldn't make the SHA256 generation work as there are type conflicts.

The following error is thrown in src/docker.rs:22:
```
return sha256.result();
           ^^^^^^^^^^^^^^^ expected struct `std::string::String`, found struct `generic_array::GenericArray`
```

Corresponding code:
```rust
fn gen_hash() -> String {
    let mut rng = rand::thread_rng();
    let mut sha256 = Sha256::new();
    sha256.input(
        rng.sample_iter(&Alphanumeric)
        .take(30)
        .collect::<String>()
    );
    return sha256.result();
}
```

I tried several things and searched for ways to convert from one type to a string but I had no luck. I also found no information about returning the hash result as a string directly.
Perhaps this is an easy fix but I am not familiar with Rust so I am stuck at the moment.

### CODE REVIEW

**Comment by svenstaro** [comment]
That was super quick! I'll have a proper look soon.

**Comment by svenstaro** [comment]
For now, it'd be good if you could fix the warnings and the errors.

**Comment by Kovah** [comment]
Done 👍

**Comment by svenstaro** [comment]
Travis still seems to be unhappy.

**Comment by svenstaro** [comment]
By the way, I don't think we really need to use any special crate here. I already generate hex numbers using `gen_hex_string()` in `utils.rs`. Isn't that good enough? We don't really need to generate proper SHAs in order for this to look right. :)

**Comment by Kovah** [comment]
I thought about this too and wanted to generate random strings but they didn't look quite like SHA hashes. But will try the gen_hex_string() function! BRB.

**Comment by svenstaro** [comment]
Feel free to add your own helper function. It doesn't need to _be_ correct, it only needs to _look_ correct. I think that's preferable to adding a whole crate for just this.

**Comment by Kovah** [comment]
Okay, refactored the hash generation to use your suggested function, it finally works! Looks good to me.

Screenshot 2019-05-22 at 13 17 22

File: data/docker_packages.txt [hunk 1]
```diff
@@ -1,50 +1,124 @@
...
+linuxserver/letsencrypt:latest
+linuxserver/letsencrypt:0.34.2-ls29
+avinetworks/se:latest
+emby/embyserver:4.2.0.6
+lsiocommunity/organizr:132
+jacobalberty/unifi:5.10.23
+skynetservices/skydns:2.5.2b
+pebbletech/cloudwatch-stats:python
+anchorfree/afmon:v1.3.0
+wurstmeister/kafka:2.12-2.2.0
```
Comment (svenstaro) [inline data/docker_packages.txt]: Wouldn't it be fun to keep the tags and the names separate and then generate them on top of each other? You'd get nonsensical stuff like `_/ubuntu:2.5-node-browsers-legacy` and that's kinda unrealistic but also fun. What do you think?

File: src/utils.rs [hunk 2]
```diff
@@ -75,6 +75,10 @@ pub fn gen_hex_string(rng: &mut ThreadRng, length: u64) -> String {
     gen_string_with_chars(rng, "0123456789abcdef", length)
 }
 
+pub fn gen_sha256_string(rng: &mut ThreadRng) -> String {
+    gen_hex_string(rng, 64)
```
Comment (svenstaro) [inline src/utils.rs]: I'm not sure specializing this call is really worth this function. Why not just always call `gen_hex_string(rng, 64)` in place? It only has a documentary effect here as it doesn't really save any repeated code. In this case, I'd actually recommend you specialize it using a lambda inside `docker.rs` like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bcccccfa7650952efa0ee65422c3cc68

I hope that gets the point across.

File: src/main.rs [hunk 3]
```diff
@@ -54,6 +55,7 @@ static COMPOSERS: &str = include_str!("../data/composer.txt");
 static SIMCITY: &str = include_str!("../data/simcity.txt");
 static BOOT_HOOKS: &str = include_str!("../data/boot_hooks.txt");
 static OS_RELEASES: &str = include_str!("../data/os_releases.txt");
+static DOCKERPACKAGES: &str = include_str!("../data/docker_packages.txt");
```
Comment (svenstaro) [inline src/main.rs]: Can you write it like `DOCKER_PACKAGES`?

**Comment by svenstaro** [comment]
Also give `clippy` a spin to show you all the lints. You can easily install it using rustup.

**Comment by svenstaro** [comment]
Great job so far, by the way. Especially considering this is your first time with Rust!

**Comment by Kovah** [comment]
Will look into your comments as soon as possible. I think the random package > tag generation would indeed be funny.

**Comment by svenstaro** [comment]
Also don't forget to update the README with this new module.

File: src/docker.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+use crate::parse_args::AppConfig;
+
+
+pub fn run(appconfig: &AppConfig) {
+    let mut rng = thread_rng();
+    let num_packages = rng.gen_range(20, 100);
+    // Choose `num_packages` packages, non-repeating and in random order
+    let chosen_packages: Vec<_> = DOCKER_PACKAGES_LIST.choose_multiple(&mut rng, num_packages).collect();
+
+    for _stage in &["Deleting"] {
```
Comment (svenstaro) [inline src/docker.rs]: What is this look supposed to do? It's looking over a list with one element. Seems like a no-op to me. Maybe some old left-over code?

File: src/docker.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+            let num_hashes = rng.gen_range(10, 30);
+            let mut index = 0;
+            while index < num_hashes {
+                println!(
+                    "Deleted: sha256:{hash}", 
+                    hash = gen_hex_string(&mut rng, 64)
+                );
+
+                index = index + 1;
```
Comment (svenstaro) [inline src/docker.rs]: You can make this into `index += 1`.

**Comment by Kovah** [comment]
Done.
Thank you for the guidance with this pull request! Really interesting small journey trough Rust. 😊

File: data/docker_packages.txt [hunk 6]
```diff
@@ -0,0 +1,124 @@
+couchbase
+redis
+redis
```
Comment (svenstaro) [inline data/docker_packages.txt]: You have lots of dups in here.

**Comment by svenstaro** [comment]
Looking great! Merging as is. Thanks for the contribution!

**Comment by Kovah** [comment]
Awesome!
Thanks again for the guidance and help with getting this done. 👍

**Pull request merged**